Write Quick Start Guide (#358)
microsoft/cobalt: docs/2_QUICK_START_GUIDE.md

# 2. Quick Start Guide

## 2.1 Overview

> *Cobalt* is a tool for developers who are interested in reusing or contributing new cloud infrastructure as code patterns in template form. An actualized infrastructure as code pattern in Cobalt is called a *Cobalt Infrastructure Template* or *CIT* (/kɪt/). Cobalt Infrastructure Templates primarily rely on [*Terraform*](https://learn.hashicorp.com/terraform)'s HCL language in order to target a wide array of cloud providers.

Reviewer: Do you also consider our Azure DevOPS templates / patterns as infrastructure templates or more PaaS pattern kit's?

Author: Per our group conversation, we are currently referring to DevOPS scripts as infrastructure as code templates.

> *Cobalt* is a tool for developers who are interested in reusing or contributing new cloud infrastructure as code patterns in template form. An actualized infrastructure as code pattern in Cobalt is called a *Cobalt Infrastructure Template* or *CIT* (/kɪt/). Cobalt Infrastructure Templates primarily rely on [*Terraform*](https://learn.hashicorp.com/terraform)'s HCL language in order to target a wide array of cloud providers.

Author: Language was updated to mention modules and a link was provided to learn more about them.

You can get pretty creative and build your own custom *CIT*s in order to use and/or contribute to Cobalt but we strongly recommend that you first complete this quick start guide. This guide is centered around our existing [*Azure Hello World CIT*](../infra/templates/az-hello-world/README.md "AZ Hello World - Cobalt Infrastructure Template") and should serve as your first Azure infrastructure deployment. In summary, completing this guide should be your first major step in familiarizing yourself with Cobalt and the *CIT* developer workflow. Happy templating! 😄

> > For a more general overview of Cobalt, please visit our main page: [READ ME](https://github.com/microsoft/cobalt/blob/master/README.md "Main Cobalt Read Me")

Reviewer: you may want to use a relative link

> > For a more general overview of Cobalt, please visit our main page: [READ ME](https://github.com/microsoft/cobalt/blob/master/README.md "Main Cobalt Read Me")

Reviewer 2: If this is not in GitHub or the repo name changes the link will break. A relative reference will fix this
```suggestion
> For a more general overview of Cobalt, please visit our main page: [READ ME](../README.md "Main Cobalt Read Me")
```

## 2.2 Goals and Objectives

🔲 Prepare local environment for *Cobalt Infrastructure Template* deployments.

🔲 Deploy the [*Azure Hello World CIT*](../infra/templates/az-hello-world/README.md "AZ Hello World - Cobalt Infrastructure Template").

> 🔲 Walk away with a introductory understanding of the *CIT* developer workflow.

Reviewer 2: I love it!

🔲 Feel confident in moving forward to our next recommended section: *[Cobalt Templating from Scratch](https://github.com/microsoft/cobalt/blob/master/docs/3_NEW_TEMPLATE.md).*

## 2.3 Prerequisites

> NOTE: Previous "infrastructure as code" experience is not a prerequisite for completing the quick start guide.

> | Azure Prereqs | Description | I | Install Prereqs | Description |

Reviewer 2: This table is confusing -- can it be broken into just 2 columns: `Prereqs` and `Description`?

Reviewer 3: I agree, this table is confusing. I'm not sure what the "I" column is intended to do. Also, this seems to be a mixture of "things you need to install first" and "things you need to go configure or enroll first". Maybe would be good to split those two categories of prereqs into different tables?

Author: Yeah, I reverted back to a single table. Keeping it simple.

|----------|--------------|-|----------|--------------|

> | `Azure Subscription` | [Azure Portal](https://portal.azure.com/) - This template needs to deploy infrastructure within an Azure subscription.|I|`Terminal with bash shell`|[WSL](https://code.visualstudio.com/docs/remote/wsl) or [Git Bash](https://git-scm.com/downloads) - The shell environment needed to follow along with the provided instructions.|

Reviewer 2: GitBash not supported - use WSL or bash

|`Azure Service Principal`|[Azure Service Principal](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal) - This template needs permissions to deploy infrastructure within an Azure subscription.|I|`Git Source Control`|[Install git](https://www.atlassian.com/git/tutorials/install-git)|

> |`Azure Storage Account`|[Azure Storage Account](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview) - An account for tracking terraform remote backend state. You can use our backend state setup [template](../infra/templates/backend-state-setup/README.md) to provision the storage resources.|I|`Terraform`|[Terraform](https://www.terraform.io/downloads.html) - Download the appropriate environment for setting up Terraform|

Reviewer 2: What version of TF (12.0+)

|`Azure CLI`|[Get started with Azure CLI](https://docs.microsoft.com/en-us/cli/azure/get-started-with-azure-cli?view=azure-cli-latest)- An account for tracking terraform remote backend state. You can use our backend state setup [template](../infra/templates/backend-state-setup/README.md) to provision the storage resources.|I|`GitHub Account`|[Github](https://github.com/login) - An account for forking the Cobalt repo|

---

## 2.4 Deploy Cobalt's [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md)

> Below are the steps for deploying the [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md). Ensure that this deployment outputs the following Azure cloud resources before you can call this quick start guide finished:

Reviewer 2: > Ensure that this deployment outputs the following Azure cloud resources before you can call this quick start guide finished

If I don't know what a terraform output is, which I may not based on the requisite knowledge called out above, then I don't understand this line. Maybe we can just call this out as a step below and omit it from here.

Author: I removed the word "output". It's a reserved term in Terraform as well. Thanks! Here's the new sentence:
> Ensure that running this template deploys the following Azure cloud resources before you can call this quick start guide finished:

| Preview | Quick Start Azure Cloud Resources |
|----------|--------------|

> |![cobalt-'NickeManarin/Screen2Gif'](./screentogif-40sec-test.gif)| <image src="https://user-images.githubusercontent.com/10041279/66533295-0fc64b80-ead8-11e9-96c0-ea5441cde854.png" width="1400"/> |

Reviewer 2: Gif seems out of place here -- does it make sense to have each step have its own step-specific gif so that it is more clear what is happening?

Author: A gif for each step would be overwhelming in my opinion. I think what matters is how the header above the gif is title in order to convey the correct context. I'll think about this a bit more. Thanks.

<!--- These gifs will have screenshots from forking, terraform plan and apply, visit azure portal and visit app service url --->

### **Step 1:** Fork Cobalt Repo

Initializing a repository that you own and control is recommended. Once forked, move onto the next step.

* From any page within this repository, find the forked menu and fork the repo into your own repository.

    ![image](https://user-images.githubusercontent.com/10041279/66366857-6e17f080-e957-11e9-8b32-266b0d4a98fc.png)

### **Step 2:** Clone Repo to Local Environment

You will be deploying Azure infrastructure for your local environment so you will need to have a copy of the Cobalt project locally.

* From any terminal with git, use the following git command to clone your forked repo into your local environment. You can find your git repo url at the landing page of your forked repo.

    ```bash
    git clone <insert-forked-git-repo-url> # ex. git clone https://github.com/<YourGitAccout>/cobalt.git
    ```

### **Step 3:** Setup Local Environment Variables

> You'll need to define a `.env` file in the root of your local project. You can use our [environment template file](https://github.com/microsoft/cobalt/blob/master/.env.template) to start. This will hold all the environment variables needed to run your Cobalt deployments locally.

Reviewer 2: Same comment r.e. relative link to the doc.

1. Navigate to the root directory of your project and use the following command to copy the environment template file.

    ```bash
    cp .env.template .env
    ```

    ```bash
    $ tree infra
    ├───.env.template
    ├───.env ## NEW FILE GENERATED FROM THE COMMAND
    ├───modules
    │   └───providers
    │       ├───azure
    │       │   ├───...
    │       │   └───vnet
    │       └───common
    └───templates
        ├───az-hello-world
        │   └───test
        └───backend-state-setup
    ```

1. Provide values for the environment values in `.env` which are required to authenticate Terraform to provision resources within your subscription.

>     ```bash

Reviewer 2: Instead of dumping it here, it may be useful to comment what these are in the `.env.template` and just add a step to fill out the values called out in the template. 

That way as the template changes, the docs don't need to be updated

    # Versioning
    GO_VERSION=1.12.5
    TF_VERSION=0.12.2
    BUILD_BUILDID=1
    # Subscription Values
    ARM_SUBSCRIPTION_ID="<az-service-principal-subscription-id>"
    ARM_CLIENT_ID="<az-service-principal-client-id>"
    ARM_CLIENT_SECRET="<az-service-principal-auth-secret>"
    ARM_TENANT_ID="<az-service-principal-tenant>"
    ARM_ACCESS_KEY="<remote-state-storage-account-primary-key>"
    # Azure Blob Storage Values
    TF_VAR_remote_state_account="<tf-remote-state-storage-account-name>"
    TF_VAR_remote_state_container="<tf-remote-state-storage-container-name>"
    TF_WARN_OUTPUT_ERRORS=1
    TF_CLI_ARGS=
    ```

1. Execute the following commands to set up the environment variables for your bash session:

    ```bash
    # these commands setup all the environment variables needed to run this template
    DOT_ENV=<path to your .env file>

>     export $(cat $DOT_ENV | xargs)

Reviewer 2: If the file has lines that start with `#`, does this work?

Author: It didn't for me. Good catch!

Reviewer 2: You may want something like this:
`export $(cat .env | grep -v '^\s*#' | xargs)`

Here is the deal w/ the grep command...
 - `^\s*` - match anything that has zero or more occurrences of a white-space character
 - `^\s*#` - match anything that has zero or more occurrences of a white-space character, followed by a `#` character -i.e., this is a comment for bash
 - `-v` - inverts the match

Author: I echoed each command to see this play out. Looks like this will work! Thanks!

    ```

1. Execute the following login command to configure your local Azure CLI.

    ```bash
    # This logs your local Azure CLI in using the configured service principal.
    az login --service-principal -u $ARM_CLIENT_ID -p $ARM_CLIENT_SECRET --tenant $ARM_TENANT_ID
    ```

    ```bash
    # Other helpful azure cli commands
    az --version # Cobalt does not recommend a version but this command proves useful for troubleshooting the az cli
    az account show # A command for validating the azure cli is correctly configured

>     az configure --defaults group='' # Use empty strings to clear default values for any sticky az properties like "group"

Reviewer 2: Is this necessary now that all the cmds use a specific RG and Sub?

Author: That's basically an opinion about an azure cli pitfall I ran into. Helpful but maybe not the right place for it. I'll remove it for now. Thanks.

Author: I left "az account show" but removed the others.

    ```

### **Step 4:** Initialize a Terraform Remote Environment

> A Terraform environment is primarily any json file generated by the `terraform init` and/or `terraform workspace new` commands. Initializing a Terraform *remote* environment (as opposed to a local one) effectively means to create an Azure blob storage json file within your existing Azure blob storage account and container listed in the [prerequisites](#23-prerequisites).

Reviewer 2: The terminology looks off here:
 - A terraform "environment" is an old way to refer to a terraform workspace (https://www.terraform.io/docs/state/environments.html) but this is not a preferred term as of now
 - The terraform syntax is `HCL` and not `json` and it is not generated by any terraform command. The workspace artifacts are, in part, JSON but there are other non JSON artifacts. Not sure its worthwhile to specify the internal workings of TF here

```suggestion
Terraform maintains a [state](https://www.terraform.io/docs/state/index.html) file that tracks important information about the resources that it has provisioned. It is a critical component to Terraform and it is worth understanding, but this will not be covered here.

A best practice when using Terraform is to store state artifacts in the cloud (i.e., Azure Blob Storage or AWS S3). In this example, we will configure a new Terraform workspace that will be persisted on Azure Blob Storage.
```

Author: I cleared this up. Less is more! Thanks!

* Navigate to the az-hello-world directory (i.e. ./infra/templates/az-hello-world) and execute the following commands to set up your remote terraform environment.

    ```bash
    # This configures terraform to leverage a remote backend that will help you and your
    # team keep consistent state
    terraform init -backend-config "storage_account_name=${TF_VAR_remote_state_account}" -backend-config "container_name=${TF_VAR_remote_state_container}"

    # This command configures terraform to use a workspace unique to you. This allows you to work
    # without stepping over your teammate's deployments

>     terraform workspace new "az-hw-$USER" || terraform workspace select "az-hw-$USER"

Reviewer 2: This will not work in GitBash FYI

    ```

    ```bash
    $ tree infra
    ├───.env.template
    ├───.env
    ├───modules
    └───templates
        ├───az-hello-world
        │   └───.terraform ## NEW LOCAL DIRECTORY GENERATED FROM INITIALIZING REMOTE STATE
        └───backend-state-setup
    ```

### **Step 5:** Deploy Cobalt's [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md)

> This is the step that demonstrates what the [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md) has to offer by deploying and creating actual Azure cloud resources. This infrastructure template creates an [ARM Resource Group](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-overview), an [App Service Plan](https://docs.microsoft.com/en-us/azure/app-service/overview-hosting-plans) and [App Service](https://docs.microsoft.com/en-us/azure/app-service/) running a public docker container. The previous step of initializing a Terraform environment from within the Azure Hello World CIT directory makes deployment possible.

Reviewer 2: What is an `ARM Resource Group`?
```suggestion
This is the step that demonstrates what the [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md) has to offer by deploying and creating actual Azure cloud resources. This infrastructure template creates an [ARM Resource Group](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-overview), an [App Service Plan](https://docs.microsoft.com/en-us/azure/app-service/overview-hosting-plans) and [App Service](https://docs.microsoft.com/en-us/azure/app-service/) running a public docker container.

You will notice that the names for the resources in Azure have been generated using the workspace name configured in the last step. This is a key component to CITs and is used by the Cobalt deployment templates to isolate resources across devint, qa and production stages. More on this in [SOME DOC?].
```

Author: Resource groups we'd see in an ARM template. I removed the "ARM" notation. Not useful here.

1. From the az-hello-world directory, execute the following commands to orchestrate a deployment.

    ```bash
    # See what terraform will try to deploy without actually deploying
    terraform plan

    # Execute a deployment
    terraform apply
    ```

1. Take note of the name of the resource group output from Terraform apply.

> * **Azure Hello World CIT [Main.tf](https://github.com/microsoft/cobalt/blob/master/infra/templates/az-hello-world/main.tf) File**

Reviewer 2: Can we omit the copy of the file? When the impl changes, it will be difficult to keep this up to date.

Also we are using a non-local ref to the template

    module "service_plan" {
        source              = "../../modules/providers/azure/service-plan"
        resource_group_name = azurerm_resource_group.main.name
        service_plan_name   = local.sp_name
    }

    module "app_service" {
        source                           = "../../modules/providers/azure/app-service"
        app_service_name_prefix          = local.app_svc_name_prefix
        service_plan_name                = module.service_plan.service_plan_name
        service_plan_resource_group_name = azurerm_resource_group.main.name
        docker_registry_server_url       = local.reg_url
        app_service_config               = local.app_services
    }
    ```

### **Step 6:** Validate Infrastructure Deployed Successfully

If you correctly completed the previous step, you can trust that Azure resources are now living in Azure. However, let's take it a step further and get your eyes on the infrastructure that was just deployed. Seeing the end result gives you the full quick start experience and is something that should be done for all future Cobalt CIT deployments.

1. Login to the Azure Portal.
1. Search for "Resource Group" to find the Resource Group menu.

> 1. Find and Select the name of the Resource Group output from the deployment.

Reviewer 2: define "output"

Author: Removed "output". It could cause confusion with Terraform template outputs and also the term is too generic like "environment".

1. Select the App Service output from the deployment.
1. Select the "overview" tab.
1. Wait for the App Service "URL" link to display itself from within the menu and then visit it.

### **Final Step:** Teardown Infrastructure

> Tearing down the infrastructure just created is as easy as running a terraform command within the scope of the same Terraform environment and workspace that created it. Gaining access to the scope of the workspace is as easy as navigating to the Cobalt CIT template directory holding a reference to the environment and workspace.

Reviewer 2: environment == workspace https://www.terraform.io/docs/state/environments.html

This description is confusing -- I'd just say that you can run these from the same directory as all the other commands.

Author: Yeah, less is more here. I agree. Here's the new summary:
> The infrastructure created from running the Azure Hello World CIT is no longer needed. It's functional but is primarily for the quick start guide. Complete the following steps to delete the resources referenced in your remote workspace:

1. From within the Azure portal, visit the Azure blob storage json file (i.e. "az-hw-$USER") which remotely holds the state of your deployed Azure infrastructure and view it's contents. The next step will wipe out the contents of this file and delete the associated Azure infrastructure.

1. Locally, from the az-hello-world directory, execute the following command to teardown your deployment and delete your resources.

    ```bash
    # Destroy resources and tear down deployment. Only do this if you want to destroy your deployment.
    terraform destroy
    ```

> 1. From within the Azure portal, revisit the Azure blob storage json file (i.e. "az-hw-$USER")  and view it's contents. At this point, this file should have an empty state.

Reviewer 2: ```suggestion
1. From within the Azure portal, revisit the Azure blob storage json file (i.e. "az-hw-$USER")  and view it's contents. At this point, this file should have an empty state. Now it can be deleted:

```bash
    # delete the workspace artifacts on blob storage - maybe this lives further down
    terraform workspace select default
    terraform workspace delete az-hw-$USER
```
```

> NOTE: Previous "infrastructure as code" experience is not a prerequisite for completing the quick start guide.

| Prereqs | Description |
|----------|--------------|
|Azure Subscription |[Azure Portal](https://portal.azure.com/) - This template needs to deploy infrastructure within an Azure subscription.|
|Azure Service Principal|[Azure Service Principal](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal) - This template needs permissions to deploy infrastructure within an Azure subscription.|
|Azure Storage Account|[Azure Storage Account](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview) - An account for tracking Terraform remote backend state. You can use our backend state setup [template](../infra/templates/backend-state-setup/README.md) to provision the storage resources.|
|Azure CLI (`Latest`)|[Get started with Azure CLI](https://docs.microsoft.com/en-us/cli/azure/get-started-with-azure-cli?view=azure-cli-latest) - The Azure CLI is a dependency of the Terraform templates used in Cobalt.|
|GitHub Account|[Github](https://github.com/login) - An account for forking the Cobalt repo.|

> |Terraform `(0.12.4)`|[Terraform](https://www.terraform.io/downloads.html) - Download the appropriate version for setting up Terraform.|

Reviewer 2: ```suggestion
|Terraform `(0.12 +)`|[Terraform](https://www.terraform.io/downloads.html) - Download the appropriate version for setting up Terraform.|
```

|Git Source Control|[Install git](https://git-scm.com/downloads).|
|Terminal with bash shell|[WSL](https://code.visualstudio.com/docs/remote/wsl) or [Bash](https://git-scm.com/downloads) - The shell environment needed to follow along with the provided instructions. *In Windows, we recommend Visual Studio Code integrated WSL*.|
---

## 2.4 Run Cobalt's [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md)

Below are the steps for running the [_Azure Hello World CIT_](../infra/templates/az-hello-world/README.md) from your terminal. Ensure that running this template achieves the quick start guide's final goal of deploying the following Azure cloud infrastructure before you can call this quick start guide finished:

| Preview Steps | Final Goal |
|----------|--------------|

> |![cobalt-'NickeManarin/Screen2Gif'](./screentogif-40sec-test.gif)| <image src="https://user-images.githubusercontent.com/10041279/66678697-149f1280-ec32-11e9-9bbb-c94a8111115b.png" width="1400"/> |

Reviewer 2: This gif has a lot of extra clicks and searches through `ctrl + r`. I'm not sure what its trying to show. Perhaps we could re-record with the steps pre-copied and pasted in another window, so it is more to the point?

Author: Gif has been updated with most steps.
